### PR TITLE
[runtime] mono_class_from_name cyclic ref handling

### PR DESCRIPTION
**Pull request by esdrubal**
The changes fix stack overflows caused while traversing images with
cyclic references.

Fixes [#31396](https://bugzilla.xamarin.com/show_bug.cgi?id=31396)

### CODE REVIEW

File: mono/metadata/class.c [hunk 1]
```diff
@@ -7822,6 +7821,38 @@ mono_class_from_name_checked (MonoImage *image, const char* name_space, const ch
...
 }
 
+MonoClass *
+mono_class_from_name_checked (MonoImage *image, const char* name_space, const char *name, MonoError *error, MonoGHashTable *visited_images)
+{
+	MonoError e;
+	MonoClass *klass = NULL;
+	gboolean owns_hash;
+
+	if (!error)
```
Comment (kumpera) [inline mono/metadata/class.c]: This is wrong.

The MonoError should be the last argument and must never be null.

File: mono/metadata/class.c [hunk 2]
```diff
@@ -7822,6 +7821,38 @@ mono_class_from_name_checked (MonoImage *image, const char* name_space, const ch
...
+	if (owns_hash)
+		visited_images = mono_g_hash_table_new (g_direct_hash, g_direct_equal);
+
+	// Checking visited images avoids visiting an image multiple times.
+	// It also avoids stack overflows when cyclic references exist.
+	if (owns_hash || !mono_g_hash_table_lookup (visited_images, image)) {
+		mono_g_hash_table_insert (visited_images, image, GUINT_TO_POINTER(1));
+
+		klass = mono_class_from_name_checked_aux (image, name_space, name, error, visited_images);
+		if (!mono_error_ok (error)) {
```
Comment (kumpera) [inline mono/metadata/class.c]: This is wrong, it's not passing the error to the caller.

File: mono/metadata/class.c [hunk 3]
```diff
@@ -7838,15 +7869,7 @@ mono_class_from_name_checked (MonoImage *image, const char* name_space, const ch
...
-	MonoError error;
-	MonoClass *klass;
-
-	klass = mono_class_from_name_checked (image, name_space, name, &error);
-	if (!mono_error_ok (&error)) {
-		mono_loader_set_error_from_mono_error (&error);
-		mono_error_cleanup (&error); /* FIXME Don't swallow the error */
-	}
-	return klass;
+	return mono_class_from_name_checked (image, name_space, name, NULL, NULL);
```
Comment (kumpera) [inline mono/metadata/class.c]: This is wrong, we should never pass a null value for MonoError.

**Comment by esdrubal** [comment]
build

**Comment by kumpera** [comment]
LGTM. Please merge if nothing shows up in jenkins.

**Pull request merged**